# Tasks
***

Attempted solutions to the Tasks assessment. Author is Shirin Nagle(G00363458@gmit.ie)
***

### Task 1: Calculate the square root of 2 correct to 100 decimal places.
***

Before I looked up how to find the square root of a  number I decided to try and see if I could come up with a way of figuring out a basic method. Which definitely means not a very efficient way of calculating a square root as I do not have a maths background. I am hoping that doing this may help me understand the problems of calculating square roots. I decided to use basic maths operators, + - / \*. To work on a small number (x) but not the number 2. To decide on what number to divide x by. 

I randomly picked 5(x) as my small number and 2(y) as my dividing number.

A square root of a number is a value that, when multiplied by itself, gives the number. Example: 4 × 4 = 16, so a square root of 16 is 4[1].



Workings:

x/y = z.

5/2 = 2.5. To check if 2.5 is the square root of 5, 2.5 * 2.5 needs to equal 5 if it is the square root.

2.5 * 2.5 = 6.25, which is not equal to 5. 
6.25 is too big, so I need to divide 5(x) by a bigger number (y)to get a smaller result(z). I added 0.5 to y. 
y = 2.5.
5/2.5 = 2. 2 * 2 = 4, which is not equal to 5, but I can say that the square root of 5 is somewhere greater than 2 and less than 2.5.

5 -(2)(2) = 1, the difference between 2 squared and 5 is 1. I decided to divide 5(x) by 2.25(y). Y is now bigger than 2 and less than 2.5.

5/2.25 = 2.22222R, 2.2222R * 2.2222R = 4.938..
2.25 * 2.25 = 5.0625

I can see that 2.222R squared is quite near to 5, and 2.25 squared is quite near to 5. So the square root must be between 2.222R and 2.25. 

5/2.23 = 2.242152466367713 * 2.242152466367713 = 5.027247682438818

The square root of 5 must be between 2.23 and 2.24. 

5/2.235 = 2.237136465324385 * 2.237136465324385 = 5.004779564484083

Getting closer to 5, the square root must be between 2.35 and 2.37

5/2.236 = 2.23613595706619 * 2.23613595706619 = 5.000304018484325



Dividing a small number by 2, 2.5 or 3 works fine for small numbers but once the numbers get bigger dividing by a small number like 2 is an extra step that may be unneccesary.

I very quickly can see how difficult it is to calulate a square root without having a method to follow.

I searched on the internet to see what the most common methods of finding a square root are.

The Babylonians are credited with possibly the first algorithm for calculating the square root of a number [2].

But decided to follow Newton's method after being pointed in the right direction.


Square root of a number can be calculated using Newton's method[3].

To find the square root of $z$ of a number $x$, we can iterate using the following equation. 

$$ z = z - \frac{z^2 - x}{2z} $$

In [1]:
def sqrt(x):
    """
    A function to calculate the square root of a number x.
    """
    # initial guess for the square root of z
    z = x / 2.5
    # loop unitl happy with accuracy 
    while abs(x - (z * z)) > 0.000001 :
        # calculate a better guess for the square root
        z -= (z*z - x) / (2 * z)
    # Return the approximate square root of x
    return z

In [2]:
# Test the function on 49
sqrt(49)

7.000000000577794

Testing the function on 2

In [4]:
sqrt(2)

1.4142135658496038

Importing math to check what the square root of 2 is.

In [48]:
import math
math.sqrt(2)

1.4142135623730951

Comparing value of the square root of 2 importing math and the value of square root of two using the written function

In [8]:
sqrt(2)

1.4142135658496038

Testing how many decimal places can be displayed.

In [4]:
def sqrt(x):
    """
    A function to calculate the square root of a number x.
    """
    # initial guess for the square root of z
    z = x / 2.5
    # loop unitl happy with accuracy 
    while abs(x - (z * z)) > 0.000001 :
        # calculate a better guess for the square root
        z -= (z*z - x) / (2 * z)
    # Return the approximate square root of x
    return z

In [5]:
sqrt(4)

2.0000000929222947

In [6]:
a = 4
sqrt(a)
print("{:.6f}".format(a))

4.000000


In [7]:
a = 4
sqrt(a)
print("{:.16f}".format(a))

4.0000000000000000


In [8]:
a = 4
sqrt(a)
print(format(a, '.12g'))

4


In [9]:
a = 4
sqrt(a)
format(a, '.12g')

'4'

In [10]:
a = 8
sqrt(a)
print(format(a, '.12g'))

8


In [11]:
a = 8
sqrt(a)


2.828427125924596

In [12]:
print(a)

8


In [13]:
a = 8
sqrt(a)
print(format(sqrt(a), '.12g'))

2.82842712592


In [14]:
a = 8
sqrt(a)
print(format(sqrt(a), '.22g'))

2.82842712592459610832


In [15]:
a = 8
sqrt(a)
print(format(sqrt(a), '.42g'))

2.828427125924596108319519771612249314785


In [16]:
a = 8
sqrt(a)
print(format(sqrt(a), '.100g'))

2.828427125924596108319519771612249314785003662109375


In [17]:
a = 8
sqrt(a)
print(format(sqrt(a), '.99g'))

2.828427125924596108319519771612249314785003662109375


In [18]:
a = 8
sqrt(a)
print(format(sqrt(a), '.2f'))

2.83


In [19]:
a = 8
sqrt(a)
print(format(sqrt(a), '.22f'))

2.8284271259245961083195


### How to display 100 Decimal places

Python shows up to 16 decimal places by default, this doesn't mean that the number being stored is limited to 16 decimal places. The number stored will contain as many decimal places as there are in the number. Looking at Python Documentation I found how to display more decimal places by using string formatting to produce a limited number of significant digits[4]
Using format(sqrt(a), '.2g') displays to 2 significant digits


In [20]:
a = 8
sqrt(a)
print(format(sqrt(a), '.2g'))

2.8


Trying the same values but showing 12 significant figures.

In [21]:
a = 8
sqrt(a)
print(format(sqrt(a), '.12g'))

2.82842712592


Trying the same values but showing 42 significant figures.

In [22]:
a = 8
sqrt(a)
print(format(sqrt(a), '.42g'))

2.828427125924596108319519771612249314785


Only 40 significant digits are displayed as the 2 digits after 85 are 0's. Trying the same values but showing 43 significant digits.

In [23]:
a = 8
sqrt(a)
print(format(sqrt(a), '.43g'))

2.828427125924596108319519771612249314785004


Trying the same values but showing 53 significant digits.

In [26]:

print(format(sqrt(a), '.53g'))

2.828427125924596108319519771612249314785003662109375


Changing the value of a to see if more digits will display

In [33]:
a = 2
sqrt(a)
print(format(sqrt(a), '.50g'))

1.414213565849603781998666818253695964813232421875


There seems to be a limit to to the digits displayed for the square root of 2, using format(a, '.50g') only displays 49 significant digits. I f I use a higer number only 49 digits are displayed, as shown below

In [34]:
a = 2
sqrt(a)
print(format(sqrt(a), '.79g'))

1.414213565849603781998666818253695964813232421875


I thought it may be possible that there were a lot of 's after ...875 but I checked on NASA's website to see what the next numbers were and it wasn't lots of 0's, NASA have calculated the square root of 2 to one million places [5].



Trying format(sqrt(a), '.2f') to see what the limit is on displaying decimal places.

In [35]:
a = 2
sqrt(a)
print(format(sqrt(a), '.2f'))

1.41


In [36]:
a = 2
sqrt(a)
print(format(sqrt(a), '.12f'))

1.414213565850


Trying to 20 decimal places

In [37]:
a = 2
sqrt(a)
print(format(sqrt(a), '.20f'))

1.41421356584960378200


Trying to 50 decimal places.

In [38]:
a = 2
sqrt(a)
print(format(sqrt(a), '.50f'))

1.41421356584960378199866681825369596481323242187500


Trying to print 99 decimal places

In [40]:
a = 2
sqrt(a)
print(format(sqrt(a), '.99f'))

1.414213565849603781998666818253695964813232421875000000000000000000000000000000000000000000000000000


The format(sqrt(a), '.99f') just seems to be adding 0's to the end of the square root.

Importing math to see what the actual output should be.

In [43]:
import math
b = math.sqrt(2)
print(b)
print(format(b,'.99f'))


1.4142135623730951
1.414213562373095145474621858738828450441360473632812500000000000000000000000000000000000000000000000


So the above looks like python is adding on 0's to make up the 99 decimal places. 

In [49]:
r = sqrt(a)
print("%.100f" % r)

1.4142135658496037819986668182536959648132324218750000000000000000000000000000000000000000000000000000


In [45]:
c = sqrt(a) * 2 ** 55
print(c)

5.095241350546061e+16


In [46]:
print(c * 10 ** 55 // 2 ** 55)

1.414213565849604e+55


 Basic familiarity with binary floating-point representation is assumed. For the above code...looking up binary floating-point representation.

Stop at any finite number of bits, and you get an approximation. On most machines today, floats are approximated using a binary fraction with the numerator using the first 53 bits starting with the most significant bit and with the denominator as a power of two. In the case of 1/10, the binary fraction is 3602879701896397 / 2 ** 55 which is close to but not exactly equal to the true value of 1/10 [4].

Testing if to the power of works without importing math

In [50]:
p = 2 ** (1/2)
print(p)

1.4142135623730951


In [55]:
print(format(p, '.50g'))
print(format(p, '.51g'))
print(format(p, '.52g'))
print(format(p, '.53g'))
print(format(p, '.54g'))

1.4142135623730951454746218587388284504413604736328
1.41421356237309514547462185873882845044136047363281
1.414213562373095145474621858738828450441360473632812
1.4142135623730951454746218587388284504413604736328125
1.4142135623730951454746218587388284504413604736328125


The square root of 2 correct to 100 places should be [5]:

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727

In [2]:
def sqrt(x):
    """
    A function to calculate the square root of a number x.
    """
    # initial guess for the square root of z
    z = x / 2.5
    # loop unitl happy with accuracy 
    while abs(x - (z * z)) > 0.000001 :
        # calculate a better guess for the square root
        z -= (z*z - x) / (2 * z)
    # Return the approximate square root of x
    return z

In [16]:
r = sqrt(2)
print(format(r, '.53g'))
repr(r)


1.414213565849603781998666818253695964813232421875


'1.4142135658496038'

### References Task 1

[1] Square Root Definition; https://www.mathsisfun.com/definitions/square-root.html

[2] Babylonian Method; https://en.wikipedia.org/wiki/Methods_of_computing_square_roots

[3] A tour of Go; Exercises: Loops and Functions;  https://tour.golang.org/flowcontrol/8

[4] Floating Point Arithmetic: Issues and Limitations; https://docs.python.org/3/tutorial/floatingpoint.html

[5] Square root of 2 to one million places; https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil

[6] Floating Point Numbers; https://floating-point-gui.de/formats/fp/

[7] IEEE 754-2008; https://en.wikipedia.org/wiki/IEEE_754-2008_revision

[8] Geeks for Geeks; https://www.geeksforgeeks.org/python-float-type-and-its-methods/


### Task 2: Chi squared Test


##### Contingency table to calculate Chi squared test for independence [1].
<br>
<span style="text-decoration: underline">                  A    B    C    D    Total</span> 
<br>
<span>White Collar 90 60    104 95 349</span>
<br>
<span>Blue Collar   30 50    51 20 151</span>
<br>
<span>No Collar     30 40   45 35  150</span>
<br>
<span style="text-decoration: bold">Total        150 150 200 150 650</span>


In [3]:
# import chi2_contingency, hard coded values for the contingency table - testing if chi2_contingency works
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html [2]
from scipy.stats import chi2_contingency
import numpy as np
obs = np.array([[90, 60, 104, 95], [30, 50, 51, 20], [30,40,45,35]])
chi2_contingency(obs)


(24.5712028585826,
 0.0004098425861096696,
 6,
 array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
        [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
        [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]]))

Worked out Chi squared value, p value and degree of freedom. These workings are contained in a folder titled Research in a document titled Task 2. The above figures of 24.5712028585826, 0.0004098425861096696, 6 and the array of figures all correspond to the values in the document titled task 2. Note the values in the document task 2 were simplified to two decimal places.


Chi squared value = 24.57, p value = 0.0004, degree of freedom = 6.
A p value of < 0.05 is thought of as being "significant" meaning we think the variables are not independent. Because 0.0004 < 0.05 we think that economic background is linked to the type of neighbourhood lived in [3].


### References Task 2

[1] Data for contingency table; Wikipedia.org; https://en.wikipedia.org/wiki/Chi-squared_test

[2] Explanation of scipy.stats; https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html

[3] Explanation of Chi squared, p value and degree of freedom with worked example; https://www.mathsisfun.com/data/chi-square-test.html


### Task 3
The standard deviation of an array of numbers x is calculated using numpy as 
np.sqrt(np.sum((x - np.mean(x))\**2)/len(x)). However, Microsoft Excel has two different versions of the standard deviation calculation, STDEV.P and STDEV.S. The STDEV.P function performs the above calculation but in the STDEV.S calculation the division is by len(x)-1rather than len(x). Research these Excel functions, writing a note in a Markdown cell about the difference between them. Then use numpy to perform a simulation demonstrating that the STDDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample. Note that part of this task is to figure out the terminology in the previous sentence.

What is Standard Deviation?
 Standard deviation is a measure of how much variance there is in a set of numbers compared to the average (mean) of the numbers. The STDEV.P function is meant to estimate standard deviation for an entire population. If data represents a sample, use the STDEV.S function[1].

In Excel STDEV.P is calculated using this fomula <br><br>
$$\begin{equation*}
\sigma =  \sqrt{\frac{\sum (x_i- \mu)^2}{N}}
\end{equation*}$$<br>
$\sigma$	= 	population standard deviation<br>
N	= 	the size of the population<br>
$x_i$	= 	each value from the population<br>
$\mu$ = the population mean. <br><br>
In Excel STDEV.S is calculated using this fomula<br><br>

s	= 	sample standard deviation<br>
N	= 	the size of the population<br>
$x_i$	= 	each value from the population<br>
$\bar{x}$ = the sample population mean. <br><br>
$$\begin{equation*}
s =  \sqrt{\frac{\sum (x_i- \bar{x})^2}{N - 1}}
\end{equation*}$$<br>
<br>



In [14]:
import numpy as np
# array of 20 numbers represents whole population
a = np.array([9, 2, 5, 4, 12, 7, 8, 11, 9, 3, 7, 4, 12, 5, 4, 10, 9, 6, 9, 4])
# standard deviation of a population
np.std(a)
print("The Standard Deviation of the population is % s" %(np.std(a)))


The Standard Deviation of the population is 2.9832867780352594


In [18]:
# https://www.geeksforgeeks.org/python-statistics-stdev/[2]
from statistics import stdev
# array of  numbers represents a sample of the whole population
sample = (9, 2, 5, 4, 12, 7)
sample1 = ( 4, 10, 9, 6, 9, 4)
sample2 = (8, 11, 9, 3, 7, 4)
sample3 = (9, 2, 5, 4, 12, 7, 8, 11, 9, 3, 7, 4)
# Print the standard deviation of   
# following sample sets of observations 
print("The Standard Deviation of Sample is % s" %(stdev(sample))) 
print("The Standard Deviation of Sample1 is % s" %(stdev(sample1))) 
print("The Standard Deviation of Sample2 is % s" %(stdev(sample2))) 
print("The Standard Deviation of Sample3 is % s" %(stdev(sample3))) 


The Standard Deviation of Sample is 3.6193922141707713
The Standard Deviation of Sample1 is 2.6832815729997477
The Standard Deviation of Sample2 is 3.03315017762062
The Standard Deviation of Sample3 is 3.194455423784382


In [1]:
# Need to wprk out STDev.S using numpy

In [ ]:
Note about sample standard deviation:[3]
    

### References Task 3
[1] What is Standard Deviation; https://exceljet.net/excel-functions/excel-stdev.p-function<br>
[2] Example of standard deviation; https://www.geeksforgeeks.org/python-statistics-stdev/<br>
[3] Difference between deviations; https://www.thoughtco.com/population-vs-sample-standard-deviations-3126372<br>

## Task 4
***
Use scikit-learn to apply k-means clustering to Fisher’s famous Iris data set. You  will  easily  obtain  a  copy  of  the  data  set  on-line.  Explain in a Markdown cell how your code works and how accurate it might be, and then explain how your model could be used to make predictions of species of iris.<br>
 The dataset is included in R base and Python in the machine learning package Scikit-learn, so that users can access it without having to find a source for it[1]. https://scikit-learn.org/stable/tutorial/basic/tutorial.html

In [1]:
from sklearn import datasets
iris = datasets.load_iris()
print(iris.data)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

Each piece of iris data is made up of 4 values, the number at index 0 relates to the Sepal Length, the number at index 1 relates to the Sepal width, the number at index 2 relates to the Petal Length, the number at index 3 relates to the Petal width.